### PR TITLE
Update build yaml pipeline matrix

### PR DESCRIPTION
**Pull request by drwill-ms**
Collapse pri1/2 jobs to singles and use parameters to determine which targets to test.
Also parameterize job parallelization and job timeout.

### CODE REVIEW

File: vsts/vsts.yaml [hunk 1]
```diff
@@ -1,5 +1,45 @@
 name: $(BuildID)_$(BuildDefinitionName)_$(SourceBranchName)_$(Date:yyyyMMdd)$(Rev:.r)
 
+# https://docs.microsoft.com/azure/devops/pipelines/process/runtime-parameters
+parameters:
```
Comment (abhipsaMisra) [inline vsts/vsts.yaml:4]: This is nice

File: vsts/vsts.yaml [hunk 2]
```diff
@@ -1,5 +1,45 @@
...
+    - netcore3.1
+    - netcore2.1
+    - net4.7.2
+    - net4.5.1
+  default: all
+
+variables:
+# Variables for the test matrix based on the 'testTargets' that turn on/off various test targets specifically or based on the 2 aggregates.
+  allTestTargets: ${{ contains(parameters['testTargets'], 'all') }}
+  minMatrix: ${{ or(or(eq(parameters['testTargets'], 'min-matrix'), eq(variables['allTestTargets'], 'True')), eq(variables['Build.Reason'], 'PullRequest')) }}
```
Comment (abhipsaMisra) [inline vsts/vsts.yaml:31]: Having not read the docs for this, I'm having some trouble interpreting this expression. Could we add a short explanation to either this file or our 1N?

File: vsts/vsts.yaml [hunk 3]
```diff
@@ -1,5 +1,45 @@
...
+
+# Minimum matrix targets
+  testNet60: ${{ or(eq(variables['minMatrix'], 'True'), contains(parameters['testTargets'], 'net6.0')) }}
+  testNetcore31: ${{ or(eq(variables['minMatrix'], 'True'), contains(parameters['testTargets'], 'netcore3.1')) }}
+
+# Remaining targets
+  testNet50: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'net5.0')) }}
+  testNetcore21: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'netcore2.1')) }}
+  testNet472: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'net4.7.2')) }}
+  testNet451: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'net4.5.1')) }}
```
Comment (abhipsaMisra) [inline vsts/vsts.yaml:41]: How does the yaml know to run allTestTargets or the minMatrix?

Comment (drwill-ms) [reply vsts/vsts.yaml:50]: Intended:

1. Someone specified it when running a build manually.
2. If a PR, it should choose min matrix.
3. If not a PR, it should choose full matrix.

I'm tweaking the logic a bit as I test.

File: vsts/determine_tests_to_run.ps1 [hunk 4]
```diff
@@ -11,14 +11,14 @@ function IsPullRequestBuild
 	return !($env:TARGET_BRANCH -and $env:TARGET_BRANCH.toLower().Contains("system.pullrequest.targetbranch"))
 }
 
-function ShouldSkipIotHubTests 
+function ShouldSkipIotHubTests
```
Comment (drwill-ms) [inline vsts/determine_tests_to_run.ps1:14]: Only removing trailing white space

File: vsts/gatedBuild.ps1 [hunk 5]
```diff
@@ -1,11 +1,19 @@
...
 # Licensed under the MIT license. See LICENSE file in the project root for full license information.
-Function IsWindows() 
+
+Function IsWindows()
 {
 	return ([Environment]::OSVersion.Platform -eq [System.PlatformID]::Win32NT)
 }
 
-if (isWindows) 
+Write-Host "SHOULD_RUN is '$($env:SHOULD_RUN)' for '$($env:FRAMEWORK)'"
```
Comment (drwill-ms) [inline vsts/gatedBuild.ps1:9]: This line is for testing only and will be removed before merging.

File: vsts/vsts.yaml [hunk 6]
```diff
@@ -17,210 +66,84 @@ resources:
     clean: true
 
 jobs:
-  ### Linux build 1 ###
-  - job: LINUX1
```
Comment (drwill-ms) [inline vsts/vsts.yaml:21]: All OS test targets collapsed into one job, rather that duplicated (DRY violation) into 2.

File: vsts/vsts.yaml [hunk 7]
```diff
@@ -17,210 +66,84 @@ resources:
...
     strategy:
-      # Change maxParallel to 1 make builds run in serial rather than in parallel
-      maxParallel: 100
+      maxParallel: ${{ parameters.maxParallelJobs }}
       matrix:
-        .Net 6.0:
+        .NET 6.0:
           FRAMEWORK: net6.0
-        .Net 5.0:
+          SHOULD_RUN: ${{ eq(variables['testNet60'], 'True') }}
```
Comment (drwill-ms) [inline vsts/vsts.yaml:78]: Create an environment variable to signal to the build script to run or not.

**Review by abhipsaMisra — APPROVED** [review]
How does /azp run know to run with minMatrix?

**Comment by drwill-ms** [comment]
> How does /azp run know to run with minMatrix?

Lines 31-40 do this. The default test target is `default`. If that is selected, then we look at whether it is a PR build and set the variables appropriately. If it is not a PR build then it runs all.

**Pull request merged**